File: performance-comparison/benchmarks/haskell-project/src/IntArith.hs
-- int_arith -- wrapping Word64 multiply-add ladder; b reads the updated a
-- (sequential).  Printed as signed 64-bit like C's %lld.
{-# LANGUAGE BangPatterns #-}
import System.Environment (getArgs)
import Data.Word (Word64)
import Data.Int (Int64)
import Data.Bits (xor)
main :: IO ()
main = do
  args <- getArgs
  let n = case args of (a:_) -> read a; _ -> 1000000 :: Int
      go :: Int -> Word64 -> Word64 -> (Word64, Word64)
      go !i !a !b
        | i >= n = (a, b)
        | otherwise =
            let a' = a * 1000003 + b
                b' = b * 999983 + a'
            in go (i + 1) a' b'
      (fa, fb) = go 0 1 1
  print (fromIntegral (fa `xor` fb) :: Int64)

File: performance-comparison/benchmarks/rust-workspace/src/bin/int_arith.rs
// int_arith -- wrapping u64 multiply-add ladder (b reads the updated a),
// xor the two accumulators, print as signed i64 bit-cast (matches C's
// %lld over int64_t and the Haskell Word64->Int64 column).
use std::env;

fn int_arith_loop(n: i64) -> i64 {
    let mut a: u64 = 1;
    let mut b: u64 = 1;
    let mut i: i64 = 0;
    while i < n {
        let new_a = a.wrapping_mul(1000003).wrapping_add(b);
        let new_b = b.wrapping_mul(999983).wrapping_add(new_a);
        a = new_a;
        b = new_b;
        i += 1;
    }
    (a ^ b) as i64
}

fn main() {
    let n: i64 = env::args()
        .nth(1)
        .and_then(|s| s.parse().ok())
        .unwrap_or(1000000);
    println!("{}", int_arith_loop(n));
}
